python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -388,6 +399,75 @@ def terminate_scheduled_nodes(self):
+        """
+        # The GCS expects Raylet ids in the request, rather than NodeProvider
+        # ids. To get the Raylet ids of the nodes to we're draining, we make
+        # the following translations of identifiers:
+        # node provider node id -> ip -> raylet id
+
+        # Convert node provider node ids to ips.
+        node_ips = {
+            self.provider.internal_ip(provider_node_id)
+            for provider_node_id in provider_node_ids_to_drain
+        }
+
+        # Only attempt to drain connected nodes, i.e. nodes with ips in
+        # LoadMetrics.
+        connected_node_ips = (
+            node_ips & self.load_metrics.raylet_id_by_ip.keys())
+
+        # Convert ips to Raylet ids.
+        raylet_ids_to_drain = {
+            self.load_metrics.raylet_id_by_ip[ip]
+            for ip in connected_node_ips
+        }
+
+        logger.info(f"Draining {len(raylet_ids_to_drain)} raylet(s).")
+        try:
+            request = gcs_service_pb2.DrainNodeRequest(drain_node_data=[
+                gcs_service_pb2.DrainNodeData(node_id=raylet_id)
+                for raylet_id in raylet_ids_to_drain
+            ])
+
+            # A successful response indicates that the GCS has marked the
+            # desired nodes as "drained." The cloud provider can then terminate
+            # the nodes without the GCS printing an error.
+            response = self.gcs_node_info_stub.DrainNode(request, timeout=5)
+
+            # Check if we succeeded in draining all of the intended nodes by
+            # looking at the RPC response.
+            drained_raylet_ids = {
+                status_item.node_id
+                for status_item in response.drain_node_status
+            }
+            failed_to_drain = raylet_ids_to_drain - drained_raylet_ids
+            if failed_to_drain:
+                self.prom_metrics.drain_node_exceptions.inc()
+                logger.error(
+                    f"Failed to drain {len(failed_to_drain)} raylet(s).")
+
+        # If we get a gRPC error with an UNIMPLEMENTED code, fail silently.
+        # This error indicates that the GCS is using Ray version < 1.8.0,
+        # for which DrainNode is not implemented.
+        except grpc.RpcError as e:
+            # If the code is UNIMPLEMENTED, pass.
+            if e.code() == grpc.StatusCode.UNIMPLEMENTED:
+                pass
+            # Otherwise, it's a plane old gRPC error and we should log it.
+            else:
+                self.prom_metrics.drain_node_exceptions.inc()
+                logger.exception(
+                    "Failed to drain Ray nodes. Traceback follows.")
+        except Exception:
+            # We don't need to interrupt the autoscaler update with an
+            # exception, but we should log what went wrong and record the
+            # failure in Prometheus.
+            self.prom_metrics.drain_node_exceptions.inc()
+            logger.exception("Failed to drain Ray nodes. Traceback follows.")
```

Comment:
You might want to throttle logging here.

Reply:
Eh, worst case, you'll get an error twice per 5 seconds. Would prefer to get the extra yelling in the logs if something is wrong.